frontends/p4/structInitializers.cpp

```diff
@@ -25,31 +181,71 @@ const IR::Expression*
 convert(const IR::Expression* expression, const IR::Type* type) {
     bool modified = false;
     if (auto st = type->to<IR::Type_StructLike>()) {
+        bool isHdr = type->is<IR::Type_Header>();
+        bool hdrInvalid = false;
         auto si = new IR::IndexedVector<IR::NamedExpression>();
         if (auto le = expression->to<IR::ListExpression>()) {
             size_t index = 0;
+            auto l = le->size();
             for (auto f : st->fields) {
+                if (index >= l) {
+                    break;
+                }
                 auto expr = le->components.at(index);
                 auto conv = convert(expr, f->type);
                 auto ne = new IR::NamedExpression(conv->srcInfo, f->name, conv);
                 si->push_back(ne);
                 index++;
             }
+            if (le->defaultInitializer) {
+                if (index == 0 && isHdr) {
+                    hdrInvalid = true;
+                }
+                const IR::Expression* expr;
+                modified = true;
+                if (!hdrInvalid) {
+                    for (; index < st->fields.size(); index++) {
+                        auto f = st->fields.at(index);
+                        expr = defaultValue(f->type, expression->srcInfo);
```

Comment:
you probably should check if this returns nullptr

Reply:
Function defaultValue() will return adequate default value for type passed as an argument, or report error and return nullptr. Why do we need nullptr check here, when error is reported and we won't use those named expressions anyway?

Reply:
You don't want the compiler to crash after reporting an error. And defensive programming can prove useful when code moves around.